### PR TITLE
Clustermeta transformation

### PR DESCRIPTION
**Pull request by kdorosh**
add new command to transformation filter to grab values in the cluster metadata.

Sample cluster metadata:
```
"metadata": {
  "filter_metadata": {
    "io.solo.transformation": {
      "host": "foo.example.com"
     }
  }
}
```

transformation filter usage:
`"{{clusterMetadata("host")}}"`
would result in:
`"foo.example.com"`

### CODE REVIEW

File: source/extensions/filters/http/transformation/inja_transformer.cc [hunk 1]
```diff
@@ -132,6 +142,78 @@ json TransformerInstance::env(const inja::Arguments& args) const {
...
 }
 
+json TransformerInstance::cluster_metadata_callback(const inja::Arguments& args) const {
+  const std::string& key = args.at(0)->get_ref<const std::string&>();
+  
+  if (!cluster_metadata_) {
+    return "";
+  }
+
+  const ProtobufWkt::Value& value = Envoy::Config::Metadata::metadataValue(cluster_metadata_, SoloHttpFilterNames::get().Transformation, key);
```
Comment (yuval-k) [inline source/extensions/filters/http/transformation/inja_transformer.cc]: i think this line wont compile. if that's the case accept the suggestion
```suggestion
  const ProtobufWkt::Value& value = Envoy::Config::Metadata::metadataValue(*cluster_metadata_, SoloHttpFilterNames::get().Transformation, key);
```

Comment (kdorosh) [reply source/extensions/filters/http/transformation/inja_transformer.cc]: They very recently changed it:
- the old version took a ref: https://github.com/envoyproxy/envoy/blob/v1.13.1/source/common/config/metadata.h#L39
- the current master (which we are now tracking) takes a pointer: https://github.com/envoyproxy/envoy/blob/master/source/common/config/metadata.h#L45

File: source/extensions/filters/http/transformation/inja_transformer.cc [hunk 2]
```diff
@@ -276,8 +358,16 @@ void InjaTransformer::transform(Http::HeaderMap &header_map,
           named_extractor.second.extract(callbacks, header_map, get_body);
     }
   }
+
+  // get cluster metadata
+  const envoy::config::core::v3::Metadata* cluster_metadata{};
+  Upstream::ClusterInfoConstSharedPtr ci = callbacks.clusterInfo();
+  if (ci.get()) {
+    cluster_metadata = &ci.get()->metadata();
```
Comment (yuval-k) [inline source/extensions/filters/http/transformation/inja_transformer.cc]:
```suggestion
    cluster_metadata = &ci->metadata();
```

File: test/extensions/filters/http/transformation/inja_transformer_test.cc [hunk 3]
```diff
@@ -565,6 +601,30 @@ TEST(InjaTransformer, ParseBodyListUsingContext) {
...
+  Http::TestHeaderMapImpl headers{{":method", "GET"}, {":path", "/foo"}};
+  TransformationTemplate transformation;
+  transformation.mutable_body()->set_text("{{clusterMetadata(\"key\")}}");
+
+  InjaTransformer transformer(transformation);
+
+  NiceMock<Http::MockStreamDecoderFilterCallbacks> callbacks;
+  
+  Upstream::MockClusterInfo* cluster_info = new NiceMock<Upstream::MockClusterInfo>();
+  Upstream::ClusterInfoConstSharedPtr cluster_info_ptr{cluster_info};
```
Comment (yuval-k) [inline test/extensions/filters/http/transformation/inja_transformer_test.cc]:
```suggestion
Upstream::ClusterInfoConstSharedPtr cluster_info_ptr(new NiceMock<Upstream::MockClusterInfo>());
```

Comment (yuval-k) [reply test/extensions/filters/http/transformation/inja_transformer_test.cc]: actually you can probably delete this block. see below

File: test/extensions/filters/http/transformation/inja_transformer_test.cc [hunk 4]
```diff
@@ -565,6 +601,30 @@ TEST(InjaTransformer, ParseBodyListUsingContext) {
...
+
+  envoy::config::core::v3::Metadata meta;
+  meta.mutable_filter_metadata()->insert({SoloHttpFilterNames::get().Transformation, MessageUtil::keyValueStruct("key", "val")});
+  ON_CALL(*cluster_info, metadata()).WillByDefault(testing::ReturnRefOfCopy(meta));
```
Comment (yuval-k) [inline test/extensions/filters/http/transformation/inja_transformer_test.cc]:
```suggestion
  ON_CALL(*callbacks.cluster_info_, metadata()).WillByDefault(testing::ReturnRefOfCopy(meta));
```
and then you can get rid of the cluster_info_ptr and the EXPECT_CALL below

File: test/extensions/filters/http/transformation/inja_transformer_test.cc [hunk 5]
```diff
@@ -565,6 +601,41 @@ TEST(InjaTransformer, ParseBodyListUsingContext) {
...
+  Http::TestHeaderMapImpl headers{{":method", "GET"}, {":path", "/foo"}};
+  TransformationTemplate transformation;
+  transformation.mutable_body()->set_text("{{clusterMetadata(\"key\")}}");
+
+  InjaTransformer transformer(transformation);
+
+  NiceMock<Http::MockStreamDecoderFilterCallbacks> callbacks;
+  envoy::config::core::v3::Metadata* meta{};
+
+  ON_CALL(*callbacks.cluster_info_, metadata()).WillByDefault(testing::ReturnRef(*meta));
```
Comment (yuval-k) [inline test/extensions/filters/http/transformation/inja_transformer_test.cc]: *meta is derefing a nil pointer which is techinically undefined behavior.
can we change to?
```
WillByDefault(testing::Return(nullptr));
```

Comment (kdorosh) [reply test/extensions/filters/http/transformation/inja_transformer_test.cc]: this won't compile, will update with similar

**Pull request merged**